cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp

```diff
@@ -0,0 +1,272 @@
+    ONEDAL_ASSERT(row_count == labels.get_dimension(0));
+    ONEDAL_ASSERT(max_inner_iter > 0);
+    ONEDAL_ASSERT(max_inner_iter <= dal::detail::limits<std::int32_t>::max());
+    ONEDAL_ASSERT(ws_count > 0);
+    ONEDAL_ASSERT(ws_count <= dal::detail::limits<std::int32_t>::max());
+    ONEDAL_ASSERT(kernel_values.get_dimension(0) == ws_count);
+    ONEDAL_ASSERT(kernel_values.get_dimension(1) == row_count);
+    ONEDAL_ASSERT(ws_indices.get_dimension(0) == ws_count);
+    ONEDAL_ASSERT(delta_alpha.get_dimension(0) == ws_count);
+    ONEDAL_ASSERT(labels.get_dimension(0) == grad.get_dimension(0));
+    ONEDAL_ASSERT(alpha.get_dimension(0) == grad.get_dimension(0));
+
+    const Float fp_min = -dal::detail::limits<Float>::max();
+
+    const Float* labels_ptr = labels.get_data();
+    const Float* kernel_values_ptr = kernel_values.get_data();
+    const std::int32_t* ws_indices_ptr = ws_indices.get_data();
+    const Float* grad_ptr = grad.get_data();
+    Float* alpha_ptr = alpha.get_mutable_data();
+    Float* delta_alpha_ptr = delta_alpha.get_mutable_data();
+    Float* grad_diff_ptr = grad_diff.get_mutable_data();
+    std::int32_t* inner_iter_count_ptr = inner_iter_count.get_mutable_data();
+
+    constexpr std::int32_t max_sg_size = 64;
+
+    const sycl::nd_range<1> nd_range = dal::backend::make_multiple_nd_range_1d(ws_count, ws_count);
+
+    auto solve_event = q.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        local_accessor_rw_t<Float> local_kernel_values(ws_count, cgh);
+        local_accessor_rw_t<Float> objective_func(ws_count, cgh);
+        local_accessor_rw_t<Float> sg_cache_values(max_sg_size + 1, cgh);
+        local_accessor_rw_t<std::int32_t> sg_cache_index(max_sg_size + 1, cgh);
+        local_accessor_rw_t<Float> local_vars(4, cgh);
+
+        cgh.parallel_for(nd_range, [=](sycl::nd_item<1> item) {
+            const std::int32_t i = item.get_local_id(0);
+
+            const std::int32_t ws_index = ws_indices_ptr[i];
+
+            Float grad_i = grad_ptr[ws_index];
+            Float alpha_i = alpha_ptr[ws_index];
+            const Float old_alpha_i = alpha_i;
+            const Float labels_i = labels_ptr[ws_index];
+
+            std::int32_t b_i = 0;
+            std::int32_t b_j = 0;
+
+            Float* local_kernel_values_ptr = local_kernel_values.get_pointer().get();
+            Float* objective_func_ptr = objective_func.get_pointer().get();
+            Float* sg_cache_values_ptr = sg_cache_values.get_pointer().get();
+            std::int32_t* sg_cache_index_ptr = sg_cache_index.get_pointer().get();
+            Float* local_vars_ptr = local_vars.get_pointer().get();
+
+            local_kernel_values_ptr[i] = kernel_values_ptr[i * row_count + ws_index];
+            item.barrier(sycl::access::fence_space::local_space);
+
+            std::int32_t inner_iter = 0;
+            for (; inner_iter < max_inner_iter; ++inner_iter) {
```

Comment:
inner_iter is int32, max_inner_iter is std::int64_t - > int overflow can be obtained - and infinite loop